From 8ebe4a149aa1538e25eed7e68893ef7f39d9131e Mon Sep 17 00:00:00 2001
From: Sathishkumar Muruganandam <murugana@qti.qualcomm.com>
Date: Tue, 24 Mar 2026 12:27:53 +0530
Subject: [PATCH] UPSTREAM:PROTOCOL: hostapd:fix invalid non-MLO TTLM element
 access

When there is per 6GHz BSS MLO to non-MLO switch with mld_ap set to 0,
during MBSSID IE addition, TTLM element MLD contexts are wrongly
accessed for non-MLO BSS case too causing hostapd seg fault.

Fix this invalid access with proper mld checks and set them to NULL
for non-MLO case.

Fixes: Idf875f4cdcab5cc21e ('hostapd: Add support for Advertised TID-to-Link Mapping')

Feature: TTLM
Patch-dependency: r83-006-UPSTREAM-PROTOCOL-hostapd-Fill-TTLM-element-in-probe-resp.patch

Signed-off-by: Sathishkumar Muruganandam <murugana@qti.qualcomm.com>
---
 src/ap/ieee802_11.c | 40 +++++++++++++++++++++++++++++++++-------
 1 file changed, 33 insertions(+), 7 deletions(-)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index ae22216847ea..5c785aabdef2 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -11186,7 +11186,12 @@ static size_t hostapd_eid_mbssid_elem_len(struct hostapd_data *hapd,
 	len = 1;
 
 	tx_xrate_len = hostapd_eid_ext_supp_rates(tx_bss, buf) - buf;
-	tx_bss_ttlm_ctx = tx_bss->mld ? &tx_bss->mld->ttlm_ctx : NULL;
+
+	/* TTLM context is only valid for MLD APs */
+	if (!tx_bss->conf->mld_ap || !tx_bss->mld)
+		tx_bss_ttlm_ctx = NULL;
+	else
+		tx_bss_ttlm_ctx = &tx_bss->mld->ttlm_ctx;
 
 	for (i = *bss_index; i < num_bss; i++) {
 		struct hostapd_data *bss;
@@ -11204,8 +11209,17 @@ static size_t hostapd_eid_mbssid_elem_len(struct hostapd_data *hapd,
 		    mbssid_known_bss(i, known_bss, known_bss_len))
 			continue;
 
-		bss_up_ttlm = &bss->mld->ttlm_ctx.upcoming_ttlm.ttlm;
-		bss_est_ttlm = &bss->mld->ttlm_ctx.established_ttlm.ttlm;
+		/*
+		 * For legacy BSSes (mld_ap = 0 or no mld struct),
+		 * do not access TTLM fields.
+		 */
+		if (!bss->conf->mld_ap || !bss->mld) {
+			bss_up_ttlm = NULL;
+			bss_est_ttlm = NULL;
+		} else {
+			bss_up_ttlm = &bss->mld->ttlm_ctx.upcoming_ttlm.ttlm;
+			bss_est_ttlm = &bss->mld->ttlm_ctx.established_ttlm.ttlm;
+		}
 
 		/*
 		 * Sublement ID: 1 octet
@@ -11276,7 +11290,8 @@ static size_t hostapd_eid_mbssid_elem_len(struct hostapd_data *hapd,
 		}
 
 		/* TTLM IE */
-		if (frame_type == WLAN_FC_STYPE_PROBE_RESP && tx_bss_ttlm_ctx) {
+		if (frame_type == WLAN_FC_STYPE_PROBE_RESP && tx_bss_ttlm_ctx &&
+			bss_up_ttlm && bss_est_ttlm) {
 			if ((tx_bss_ttlm_ctx->established_ttlm.ttlm.expected_duration_present ||
 			     tx_bss_ttlm_ctx->upcoming_ttlm.ttlm.mapping_switch_time_present) ||
 			    (bss_est_ttlm->expected_duration_present ||
@@ -11436,8 +11451,18 @@ static u8 * hostapd_eid_mbssid_elem(struct hostapd_data *hapd, u8 *eid, u8 *end,
 		    mbssid_known_bss(i, known_bss, known_bss_len))
 			continue;
 		conf = bss->conf;
-		bss_up_ttlm = &bss->mld->ttlm_ctx.upcoming_ttlm.ttlm;
-		bss_est_ttlm = &bss->mld->ttlm_ctx.established_ttlm.ttlm;
+
+		/*
+		 * For legacy BSSes (mld_ap = 0 or no mld struct),
+		 * do not access TTLM fields.
+		 */
+		if (!bss->conf->mld_ap || !bss->mld) {
+			bss_up_ttlm = NULL;
+			bss_est_ttlm = NULL;
+		} else {
+			bss_up_ttlm = &bss->mld->ttlm_ctx.upcoming_ttlm.ttlm;
+			bss_est_ttlm = &bss->mld->ttlm_ctx.established_ttlm.ttlm;
+		}
 
 		*eid++ = WLAN_MBSSID_SUBELEMENT_NONTRANSMITTED_BSSID_PROFILE;
 		eid_len_pos = eid++;
@@ -11511,7 +11536,8 @@ static u8 * hostapd_eid_mbssid_elem(struct hostapd_data *hapd, u8 *eid, u8 *end,
 		eid += xrate_len;
 
 		/* TTLM IE */
-		if (frame_type == WLAN_FC_STYPE_PROBE_RESP && tx_bss_ttlm_ctx) {
+		if (frame_type == WLAN_FC_STYPE_PROBE_RESP && tx_bss_ttlm_ctx &&
+			bss_up_ttlm && bss_est_ttlm) {
 			/* if tx bss has non-default ttlm mapping, include ttlm element for non-tx
 			 * bss to avoid TTLM element inheritance. if non-tx vap does not have any
 			 * non-default mapping advertised, default ttlm element is included in this
-- 
2.34.1

